### DISCUSSION TITLE
Module file direct download - from TAS

### DISCUSSION THREAD

**Discussion by btsimonh** [tasmota/Berry_playground#19] [Playground Management]
I was thinking about our modules and making them easy to use in TAS.
Struck me that Berry can download them, and store them in the FS.
So I created a JSON file in my fork, and use the below code to request a named module and dependencies....

Question:
I'd like modules to be in folders in the flash FS, but I can't find a way to make a folder in Berry.  Tas excludes the OS module, which is where mkdir() resides....  Does anyone know of a way?

If we feel this would be a useful feature of this repo, then we'd need to define the JSON, and determine where THIS code would live :)

```
import json
#import os

var rootsimon = "https://raw.githubusercontent.com/btsimonh/Berry_playground/grabber"

class playgroundModules
  var root
  var files
  def init(r)
    self.root = r
    self.files = nil
  end
  
  def getFile(path)
    var cl = webclient()
    cl.begin(root .. path['src'])
    var res = cl.GET()
    print(res)
    if res == 200
      print('writing '..path['dest'])
      var s = cl.write_file(path['dest'])
      print(s)
    else
      print('failed '..path['dest'])
    end
  end

  def readModules()
    var cl = webclient()
    cl.begin(self.root .. "/api/files.json")
    var res = cl.GET()
    print(res)
    if res == 200
      var s = cl.get_string()
      self.files = json.load(s)
    end
  end
  
  def getModule(name)
    print('getmodule '..name)
    if !self.files
      self.readModules()
    end
    print(self.files)
    if !self.files
      return false
    end
    
    print(self.files['modules'])
    for file:self.files['modules']
      if file['name'] == name
        print('found '..name)
        if file.contains('dependencies')
          for dep:file['dependencies']
            print('depends on '..dep)
            self.getModule(dep)
          end
        end
        
        if file.contains('files')
          print(file)
          for f:file['files']
            print(f)
            self.getFile(f)
          end
        end
      end
    end
  end
end

var m = playgroundModules(rootsimon)
m.getModule('googledrive')
```

**Comment by Jason2866** [comment]
There is no folders support

**Reply by btsimonh** [reply]
hmm.  If I add path.mkdir to berry, it seems to work.  The manage file system display looks a little odd, but could be fixed.
Looks like mkdir and rmdir have been used from Scripting.  Is there a reason why we would not want them (mkdir, rmdir) exposed?

**Reply by Jason2866** [reply]
One reason, no code add wanted for a not needed feature. There are zip (no compression) supported.
A PR will not be accepted, since a lot of changes ahead on several (to many places) by doing this

**Reply by btsimonh** [reply]
maybe not so hasty - the changes and code addition are minimal for berry to support folders. (2 files, 1 function, 2 proxy functions).
Let me play and report.  Agreed, it may not be a needed feature - until you get to using 32Gbyte SD cards.
For local storage of timelapse images, I'd like to have them in a folder inside the SD.
For berry modules, a folder structure is attractive.

**Reply by Jason2866** [reply]
FS function should work same for esp8266 and esp32. Every byte counts for esp8266.
There is no problem to have many files without folders.

**Comment by btsimonh** [comment]
In TAS or just in the Berry implementation?

**Comment by Jason2866** [comment]
Both

**Comment by dkebler** [comment]
for now modules could be bundled as a tapp file which at least avoids the clutter of many individual files in the root. 

Also given this "dev" setup I have (that isn't available, yet) it can already can send download requests to a tasmota device (via the api).  So the dev "environment" can  initiate pulling the needed modules from the playground to the dev machine (if desired) and/or directly to a tasmota device.  In essence we are kinda making a npm like repository of the playground.   A berry "project" could even have something like a package.json to indicate which modules are required  (but please not json so it can have comments and be human edited...yaml?).   

I'll start a discussion on my dev environment code.   Also, I see no reason it can't be bundled into a vscode extension since it's written in node/javascript.  This which will make deployment for other easier.

**Reply by btsimonh** [reply]
ahaa!  Yes, exactly the type of thing I was thinking - I have no knowledge (yet) of tapp or compiled Berry files... but some formal way of easily accessing the modules provided here without messing about with downlaoding them, then uploading them, etc.
P.s. we do now have folder support in the development branch of TAS....  so it's now possible to path.mkdir('/myfolder') and path.rmdir('/somefolderthatisempty'), and I've already coded a zap('/somefolderwhichisnotempty') in berry....

I'm hoping for static file serving too (PR is in) - which would enable us to serve HTML/JS direct from the tas filesystem, enabling much processing to be done in browser, possibly using an external (e.g. github hosted) app.  That's what I created for OpenBeken, and it's worked quite well...  (I have an alteria motive - wanting to serve saved jpegs directly from TAS).

**Reply by btsimonh** [reply]
does Berry have JSONP?  hate Yaml....

**Comment by Jason2866** [comment]
> I'm hoping for static file serving too (PR is in) - which would enable us to serve HTML/JS direct from the tas filesystem, enabling much processing to be done in browser, possibly using an external (e.g. github hosted) app. That's what I created for OpenBeken, and it's worked quite well... (I have an alteria motive - wanting to serve saved jpegs directly from TAS).

This is already possible.

**Reply by btsimonh** [reply]
How do I enable it?  That part of the PR may be redundant...

**Reply by btsimonh** [reply]
@Jason2866 - i've searched and can't find a reference.  please point me in the right direction?
